controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/AbstractRequestProcessor.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    public CompletableFuture<Void> processTruncateStream(TruncateStreamEvent truncateStreamEvent) {
+        return Futures.failedFuture(new RequestUnsupportedException("Request Unsupported"));
+    }
+
+    @Override
+    public CompletableFuture<Void> processSealStream(SealStreamEvent sealStreamEvent) {
+        return Futures.failedFuture(new RequestUnsupportedException("Request Unsupported"));
+    }
+
+    @Override
+    public CompletableFuture<Void> processDeleteStream(DeleteStreamEvent deleteStreamEvent) {
+        return Futures.failedFuture(new RequestUnsupportedException("Request Unsupported"));
+    }
+
+    protected <T extends ControllerEvent> CompletableFuture<Void> withCompletion(StreamTask<T> task, T event, String scope, String stream,
+                                                                                 Predicate<Throwable> writeBackPredicate) {
+        Preconditions.checkNotNull(task);
+        Preconditions.checkNotNull(event);
+        Preconditions.checkNotNull(scope);
+        Preconditions.checkNotNull(stream);
+        Preconditions.checkNotNull(writeBackPredicate);
+        CompletableFuture<Void> resultFuture = new CompletableFuture<>();
+
+        OperationContext context = streamMetadataStore.createContext(scope, stream);
+        suppressException(streamMetadataStore.getWaitingRequestProcessor(scope, stream, context, executor), null,
+                "Exception while trying to fetch waiting request. Logged and ignored.")
+                .thenAccept(waitingRequestProcessor -> {
+                    if (waitingRequestProcessor == null || waitingRequestProcessor.equals(getProcessorName())) {
+                        withRetries(() -> task.execute(event), executor)
+                                .whenComplete((r, ex) -> {
+                                    if (ex != null && writeBackPredicate.test(ex)) {
+                                        suppressException(streamMetadataStore.createWaitingRequestIfAbsent(scope, stream, getProcessorName(), context, executor),
+                                                null, "Exception while trying to create waiting request. Logged and ignored.")
+                                                .thenCompose(ignore ->  retryIndefinitelyThenComplete(() -> task.writeBack(event), resultFuture, ex));
+                                    } else {
+                                        // Processing was done for this event, whether it succeeded or failed, we should remove
+                                        // the waiting request if it matches the current processor.
+                                        // If we dont delete it then some other processor will never be able to do the work.
+                                        // So we need to retry indefinitely until deleted.
+                                        retryIndefinitelyThenComplete(() -> streamMetadataStore.deleteWaitingRequestConditionally(scope,
```

Comment:
This is not an ephemeral node. It would work just as well if it were ephemeral too. 
And this node gets cleaned up at the end of each processing whether it succeeded or failed. 

If a controller process crashes before cleaning this up, the checkpoint on the "stream" is guaranteed to be behind or equal to position of current event. So upon recovery this event is processed again and irrespective of the result of processing of the event, upon completion the "waiting proc" node is cleaned up.

Secondly, this "working processor" node is per stream. So if there is a working node which we are unable to delete, processing for that stream will be stalled and not across all streams or all controller instances. 

Thirdly, we can have a "count" for number of times processing was ignored or instead of introducing more state we can make it probabilistic. if the flag is set, based on some probability ignore or honor it.  what do you think?

Reply:
Frankly, I'm a bit lost here. On the one hand, I'm worried about a processor taking a long time to remove the waiting flag (e.g., long GC pause, crashed and took time to recover). On the other, I wouldn't know how to set a count or even a probability given that I don't have enough insight on the overall processing times. We were talking the other day about one second to commit txns, which sounded like a lot, so I'm not sure what the right path is.

Reply:
I am not sure i follow you here. 
If you are concerned that if a processing completes on processor 1 but it fails to delete the `fairness barrier`(waiting node) and crashes then all other processing on "other" processors is stalled until processor 1 recovers and eventually deletes the node. This could take `recovery-delay` + `idempotent processing time` + `time to delete znode`. 

This can be mitigated by employing a probabilistic honoring of `fairness barrier`. It is important to understand that not honoring fairness barrier is not a correctness issue. Each processing takes a distributed lock (changing the state) which is used for mutex.